colexec: flake in TestRandomizedCast

Found here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/8972190?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
=== RUN   TestRandomizedCast
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/e8ecb0a2d1daf60d35be0846c86e1f60/logTestRandomizedCast609659391
    test_log_scope.go:79: use -show-logs to present logs inline
panic: test timed out after 4m55s
goroutine 49 [running]:
testing.(*M).startAlarm.func1()
  GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
  GOROOT/src/time/sleep.go:176 +0x32
goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0xc000fb96c0, {0x2e1d77f?, 0x610dc5?}, 0x2fae1a8)
  GOROOT/src/testing/testing.go:1494 +0x37a
testing.runTests.func1(0xc001adb8c0?)
  GOROOT/src/testing/testing.go:1846 +0x6e
testing.tRunner(0xc000fb96c0, 0xc000d1ce60)
  GOROOT/src/testing/testing.go:1446 +0x10b
testing.runTests(0xc000432b40?, {0x5344880, 0x6, 0x6}, {0x1?, 0x23?, 0x5633080?})
  GOROOT/src/testing/testing.go:1844 +0x456
testing.(*M).Run(0xc000432b40)
  GOROOT/src/testing/testing.go:1726 +0x5d9
pkg/sql/colexec/colexecbase/colexecbase_test_test.TestMain.func1(0x2e41d6f?)
  pkg/sql/colexec/colexecbase/colexecbase_test_test/pkg/sql/colexec/colexecbase/main_test.go:81 +0x65f
pkg/sql/colexec/colexecbase/colexecbase_test_test.TestMain(0x2e41d6f?)
  pkg/sql/colexec/colexecbase/colexecbase_test_test/pkg/sql/colexec/colexecbase/main_test.go:82 +0x28
main.main()
  main/bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/colexecbase_test_/testmain.go:123 +0x472
goroutine 59 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
  github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:75 +0x5a
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
  github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:41 +0x25
goroutine 60 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
  github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:98 +0xdc
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
  github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:42 +0x31
goroutine 46 [syscall, 4 minutes]:
os/signal.signal_recv()
  GOROOT/src/runtime/sigqueue.go:152 +0x2f
os/signal.loop()
  GOROOT/src/os/signal/signal_unix.go:23 +0x19
created by os/signal.Notify.func1.1
  GOROOT/src/os/signal/signal.go:151 +0x2a
goroutine 47 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0.func1()
  github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:165 +0x125
created by github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0
  github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:157 +0x25
goroutine 63 [runnable]:
math/rand.New(...)
  GOROOT/src/math/rand/rand.go:67
github.com/cockroachdb/cockroach/pkg/util/randutil.NewTestRand()
  github.com/cockroachdb/cockroach/pkg/util/randutil/rand.go:85 +0x245
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.(*opTestInput).Next(0xc000786630)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:987 +0x1bd
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next(0xc001b400a0)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/operator.go:135 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*castDatumStringOp).Next(0xc000308380)
  github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/cast.eg.go:13988 +0x3f
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.(*OpTestOutput).next(0xc000e60000)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:1277 +0x5d
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.(*OpTestOutput).Verify.func1()
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:1306 +0x26
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x2aa7ce0?)
  github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.(*OpTestOutput).Verify(0xc000e60000)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:1305 +0x9f
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.RunTestsWithoutAllNullsInjectionWithErrorHandler.func2(0xc000fb9860, {0xc000e4e8a0?, 0xc001daa000?, 0x1d24?})
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:533 +0x1da
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.RunTestsWithFn(0x10?, 0x7f1ba74eefff?, {0xc001a19858, 0x1, 0x1?}, {0xc001a19840, 0x1, 0x442d33?}, 0xc001a19030)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:711 +0x20a
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.RunTestsWithoutAllNullsInjectionWithErrorHandler(0xc000fb9860, 0x8?, {0xc001a19858?, 0x1, 0x0?}, {0xc001a19840?, 0x1, 0x0?}, {0xc001de0000, 0x1d24, ...}, ...)
  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils/utils.go:523 +0x1a9
pkg/sql/colexec/colexecbase/colexecbase_test_test.TestRandomizedCast(0xc000fb9860)
  pkg/sql/colexec/colexecbase/colexecbase_test_test/pkg/sql/colexec/colexecbase/cast_test.go:82 +0x505
testing.tRunner(0xc000fb9860, 0x2fae1a8)
  GOROOT/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
  GOROOT/src/testing/testing.go:1493 +0x35f
goroutine 64 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc001b2d040, {0x3bbc890, 0xc00002ddc0})
  github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x7f
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
  github.com/cockroachdb/cockroach/pkg/util/log/flags.go:306 +0x112a
goroutine 65 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc001b2d110, {0x3bbc890, 0xc00002ddc0})
  github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x7f
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
  github.com/cockroachdb/cockroach/pkg/util/log/flags.go:306 +0x112a
goroutine 82 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc001b2d1e0, {0x3bbc890, 0xc00002ddc0})
  github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x7f
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
  github.com/cockroachdb/cockroach/pkg/util/log/flags.go:306 +0x112a
goroutine 83 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc001b2d2b0, {0x3bbc890, 0xc00002ddc0})
  github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x7f
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
  github.com/cockroachdb/cockroach/pkg/util/log/flags.go:306 +0x112a
```

Jira issue: CRDB-25133

--- a/pkg/sql/colexec/colexecbase/cast.eg.go
+++ b/pkg/sql/colexec/colexecbase/cast.eg.go
@@ -1242,6 +1242,16 @@ type castIdentityOp struct {
 
 var _ colexecop.ClosableOperator = &castIdentityOp{}
 
+// identityOrder is a slice in which every integer equals its ordinal.
+var identityOrder []int
+
+func init() {
+	identityOrder = make([]int, coldata.BatchSize())
+	for i := range identityOrder {
+		identityOrder[i] = i
+	}
+}
+
 func (c *castIdentityOp) Next() coldata.Batch {
 	batch := c.Input.Next()
 	n := batch.Length()
@@ -1250,17 +1260,19 @@ func (c *castIdentityOp) Next() coldata.Batch {
 	}
 	projVec := batch.ColVec(c.outputIdx)
 	c.allocator.PerformOperation([]coldata.Vec{projVec}, func() {
-		maxIdx := n
+		srcVec := batch.ColVec(c.colIdx)
 		if sel := batch.Selection(); sel != nil {
 			// We don't want to perform the deselection during copying, so we
-			// will copy everything up to (and including) the last selected
-			// element, without the selection vector.
-			maxIdx = sel[n-1] + 1
+			// use a special copy in which we use the identity order but apply
+			// the selection vector.
+			projVec.CopyWithReorderedSource(srcVec, sel[:n], identityOrder)
+		} else {
+			projVec.Copy(coldata.SliceArgs{
+				Src:         srcVec,
+				SrcStartIdx: 0,
+				SrcEndIdx:   n,
+			})
 		}
-		projVec.Copy(coldata.SliceArgs{
-			Src:       batch.ColVec(c.colIdx),
-			SrcEndIdx: maxIdx,
-		})
 	})
 	return batch
 }

--- a/pkg/sql/colexec/colexecbase/cast_tmpl.go
+++ b/pkg/sql/colexec/colexecbase/cast_tmpl.go
@@ -310,6 +310,16 @@ type castIdentityOp struct {
 
 var _ colexecop.ClosableOperator = &castIdentityOp{}
 
+// identityOrder is a slice in which every integer equals its ordinal.
+var identityOrder []int
+
+func init() {
+	identityOrder = make([]int, coldata.BatchSize())
+	for i := range identityOrder {
+		identityOrder[i] = i
+	}
+}
+
 func (c *castIdentityOp) Next() coldata.Batch {
 	batch := c.Input.Next()
 	n := batch.Length()
@@ -318,17 +328,19 @@ func (c *castIdentityOp) Next() coldata.Batch {
 	}
 	projVec := batch.ColVec(c.outputIdx)
 	c.allocator.PerformOperation([]coldata.Vec{projVec}, func() {
-		maxIdx := n
+		srcVec := batch.ColVec(c.colIdx)
 		if sel := batch.Selection(); sel != nil {
 			// We don't want to perform the deselection during copying, so we
-			// will copy everything up to (and including) the last selected
-			// element, without the selection vector.
-			maxIdx = sel[n-1] + 1
+			// use a special copy in which we use the identity order but apply
+			// the selection vector.
+			projVec.CopyWithReorderedSource(srcVec, sel[:n], identityOrder)
+		} else {
+			projVec.Copy(coldata.SliceArgs{
+				Src:         srcVec,
+				SrcStartIdx: 0,
+				SrcEndIdx:   n,
+			})
 		}
-		projVec.Copy(coldata.SliceArgs{
-			Src:       batch.ColVec(c.colIdx),
-			SrcEndIdx: maxIdx,
-		})
 	})
 	return batch
 }

